File: src/minimizeDFA.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;
import javafx.util.Pair;


public class minimizeDFA {
    public static void main(String[] args) throws FileNotFoundException {
        // Section 1
        List<Integer> edges = new ArrayList<>();
        List<StateNode> states = new ArrayList<>();
        Scanner in = new Scanner(System.in);
        System.out.println("Provide a dfa file to minimize: ");
        String userInput = in.nextLine();
        File file = new File("../" + userInput);
        Scanner f = new Scanner(file);
        f.next(); // removes top left mark
        for(String s : f.nextLine().split(" ")){
            edges.add(Integer.parseInt(s));
        }
        while (f.hasNextLine()){
            String[] arr = f.nextLine().split(" ");
            int state = Integer.parseInt(arr[0]);
            StateNode thisNode = new StateNode(state);
            for(int i = 1; i < arr.length; i++){
                thisNode.children.put(edges.get(i-1), Integer.parseInt(arr[i]));
            }
            states.add(thisNode);
        }

        // Section 2

        List<Pair> grid = new ArrayList<>();
        List<Pair> marked = new ArrayList<>();
        for(StateNode node : states){
            int x = node.val;
            for(StateNode b : states){
                int y = b.val;
                if(x != y && !grid.contains(new Pair(String.valueOf(x),String.valueOf(y)))){
                    grid.add(new Pair(String.valueOf(x),String.valueOf(y)));
                    if((x.contains('F') && !y.contains('F'))||(!x.contains('F') && y.contains('F'))){
                        marked.add(new Pair(String.valueOf(x),String.valueOf(y)));
                    }
                }
            }
        }
        // Section 3
        for(int i = 0; i < grid.size() - marked.size(); i++){
            for(Pair p : grid){
                if(!marked.contains(p) && p.size > 0){
                    Pair temp = new Pair(p.x, p.y);
                    String oldx = temp.x;
                    String oldy = temp.y;
                    for(int edge : edges){
                        int newy = states.get(Integer.parseInt(String.valueOf(oldy.charAt(0)))).children.get(edge);
                        int newx = states.get(Integer.parseInt(String.valueOf(oldx.charAt(0)))).children.get(edge);

                        if(marked.contains(new Pair(String.valueOf(newx), String.valueOf(newy)))){
                            marked.add(p);
                        }
                    }

                }
            }
        }

        // Section 4

        marked = new ArrayList<Pair>(new HashSet<>(marked));
        for(Pair x : marked){
            grid.remove(x);
        }
        //Section 5

        Map<Object, Object> toDelete = new HashMap<>();
        for(Pair p : grid){
            Pair temp = new Pair(p.x, p.y);
            String first = temp.x;
            String second = temp.y;
            int i = Integer.parseInt(second);
            int j = Integer.parseInt(first);
        }
    }

    private static class StateNode {
        int val;
        Map<Integer, Integer> children;

        public StateNode(int val) {
            this.val = val;
            children = new HashMap<>();
        }

        public StateNode(int val, Map<Integer, Integer> children) {
            this.val = val;
            this.children = children;
        }
    }

    private static class Pair{
        final int size = 2;
        String x;
        String y;

        public Pair(String x, String y) {
            this.x = x;
            this.y = y;
        }
    }

}

File: MinimizeDFA.py
from typing import Set

from StateNode import StateNode as sn

edges = []
states = []
userInput = input()
with open(str(userInput)) as file:
    firstLine = file.readline().split()
    for i in range(1, len(firstLine)):
        edges.append(firstLine[i])
    # populate the map
    line = file.readline()
    while line:
        arr = line.split()
        state = arr[0]
        thisNode = sn(state)
        for i in range(1, len(arr)):
            thisNode.children[edges[i - 1]] = arr[i]
        states.append(thisNode)
        line = file.readline()

grid = []
marked = []
for a in states:
    x = a.val
    for b in states:
        y = b.val
        if x is not y and {x, y} not in grid:
            grid.append({x, y})
            if ('F' in x and 'F' not in y) or ('F' not in x and 'F' in y):
                marked.append({x, y})

# grid populated and first round of marking is done

for i in range(len(grid) - len(marked)):
    for combo in reversed(grid):
        if combo not in marked and len(combo) > 0:
            temp = combo.copy()
            oldY = temp.pop()
            oldX = temp.pop()
            for edge in edges:
                newY = states[int(oldY[0])].children[edge]
                newX = states[int(oldX[0])].children[edge]

                if {newX, newY} in marked:
                    marked.append(combo)

# finding all non marked states
# pop all of the unnecessary states
for x in marked:
    grid.remove(x)
# combine all left

# remove only the repeats
toDelete = dict()
for combo in grid:
    temp = combo.copy()
    second = temp.pop()
    first = temp.pop()
    i = int(second[0])
    j = int(first[0])

    # gross? combining all of the key pairs into a single map that maps all of the to delete ones to a single state to
    # keep and delete the rest
    if j not in toDelete.keys() and i not in toDelete.keys():
        if j not in toDelete.values() and i not in toDelete.values():
            toDelete[i] = set()
            toDelete[i].add(j)
        elif j in toDelete.values():
            for a in toDelete.keys():
                if j in toDelete[a]:
                    toDelete[a].add(i)
        elif i in toDelete.values():
            for a in toDelete.keys():
                if i in toDelete[a]:
                    toDelete[a].add(j)
    if i in toDelete.keys() and j in toDelete.keys():
        toDelete[i].add(x for x in toDelete[j])
        toDelete[i].add(j)
        toDelete.pop(j)
    elif i in toDelete.keys():
        toDelete[i].add(j)
    else:
        toDelete[j].add(i)


deleteStates = set()
for x in toDelete.keys():
    for y in toDelete[x]:
        if isinstance(y, int):
            deleteStates.add(y)
deleteStates = list(deleteStates)
deleteStates.sort()
for i in reversed(deleteStates):
    for edge in edges:
        stc = states[i].children[edge]
        if int(stc[0]) in deleteStates:
            for x in toDelete.keys():
                if i in toDelete[x]:
                    states[i].children[edge] = states[x].val
    del states[i]

print("             ", end='')
for x in edges:
    print(edge + "   ", end='')
print()
for x in states:
    print("state: " + x.val, end='')
    print(" " * (2 - len(x.val)) + " | ", end='')
    for y in x.children.values():
        print(y + " " * (4 - len(y)), end='')
    print()

print('Where: ')
for x in toDelete.keys():
    print(str(x) + " maps to ", end='')
    for y in toDelete[x]:
        if isinstance(y, int):
            print(str(y) + " : ", end='')
    print()


# regenerate DFA Table
